s3stream/src/main/java/com/automq/stream/s3/operator/DefaultS3Operator.java

```diff
@@ -442,24 +444,73 @@ public CompletableFuture<Void> delete(String path) {
         ObjectIdentifier[] toDeleteKeys = objectKeys.stream().map(key ->
-            ObjectIdentifier.builder()
-                .key(key)
-                .build()
+                ObjectIdentifier.builder()
+                        .key(key)
+                        .build()
         ).toArray(ObjectIdentifier[]::new);
+
+        // we need quiet here because BOS S3 API works as quiet mode
+        // in this mode success delete objects won't be returned.
+        // which could cause object not deleted in metadata.
+        //
+        // BOS doc: https://cloud.baidu.com/doc/BOS/s/tkc5twspg
+        // S3 doc: https://docs.aws.amazon.com/AmazonS3/latest/API/API_DeleteObjects.html#API_DeleteObjects_RequestBody
+
         DeleteObjectsRequest request = DeleteObjectsRequest.builder()
-            .bucket(bucket)
-            .delete(Delete.builder().objects(toDeleteKeys).build())
-            .build();
-        // TODO: handle not exist object, should we regard it as deleted or ignore it.
-        return this.writeS3Client.deleteObjects(request).thenApply(resp -> {
-            S3OperationStats.getInstance().deleteObjectsStats(true).record(timerUtil.elapsedAs(TimeUnit.NANOSECONDS));
-            LOGGER.info("[ControllerS3Operator]: Delete objects finished, count: {}, cost: {}", resp.deleted().size(), timerUtil.elapsedAs(TimeUnit.NANOSECONDS));
-            return resp.deleted().stream().map(DeletedObject::key).collect(Collectors.toList());
-        }).exceptionally(ex -> {
-            S3OperationStats.getInstance().deleteObjectsStats(false).record(timerUtil.elapsedAs(TimeUnit.NANOSECONDS));
-            LOGGER.info("[ControllerS3Operator]: Delete objects failed, count: {}, cost: {}, ex: {}", objectKeys.size(), timerUtil.elapsedAs(TimeUnit.NANOSECONDS), ex.getMessage());
-            return Collections.emptyList();
-        });
+                .bucket(bucket)
+                .delete(Delete.builder()
+                        .objects(toDeleteKeys)
+                        .quiet(true)
+                        .build())
+                .build();
+
+        return this.writeS3Client.deleteObjects(request)
+                .thenApply(resp -> handleDeleteObjectsResponse(objectKeys, timerUtil, resp))
+                .exceptionally(ex -> {
+                    S3OperationStats.getInstance().deleteObjectsStats(false)
+                            .record(timerUtil.elapsedAs(TimeUnit.NANOSECONDS));
+                    LOGGER.info("[ControllerS3Operator]: Delete objects failed, count: {}, cost: {}, ex: {}",
+                            objectKeys.size(), timerUtil.elapsedAs(TimeUnit.NANOSECONDS), ex.getMessage());
+                    return Collections.emptyList();
+                });
+    }
+
+    static List<String> handleDeleteObjectsResponse(List<String> objectKeys,
+                                                    TimerUtil timerUtil,
+                                                    DeleteObjectsResponse response) {
```

Comment:
The response code is not 200, did the S3Client return it through thenApply or exceptionally?

If it is returned through thenApply, then it is also necessary to judge the code of the response.